mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td

```diff
@@ -680,6 +680,125 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
     }];
 }
 
+//===----------------------------------------------------------------------===//
+// Contract op.
+//===----------------------------------------------------------------------===//
+
+def ContractOp : LinalgStructuredBase_Op<"contract", [
+               AttrSizedOperandSegments,
+               LinalgContractionOpInterface]> {
+  let summary = [{
+    Perform a contraction on two inputs, accumulating on top of a third.
+  }];
+  let description = [{
+    The semantics of contracting inputs `A` and `B` on top of `C` to produce
+    output `D` is given by
+
+      `D[H] = (SUM_{(I ∪ J) \ H} A[I] * B[J]) + C[H]`
```

Comment:
This is hard-coding the actual operation kind as summation, i.e. `+`. Apologies if I missed this somewhere, but shouldn't this be an arbitrary operation? Or is this PR meant to be limited to one kind?

Reply:
Indeed, I ended up going with the minimal version after @ftynse [pointed out](https://discourse.llvm.org/t/mlir-rfc-introduce-linalg-contract/83589/7) that only varying one operation is mathematically suspect. Note that `vector.contract` does only allow control over the accumulating operation.

I feel we probably should first determine if want to allow independent control over the add and mul operators or if these should be specified as a pair for which the verifier checks that they belong to a (known) field.

In any case, there's a TODO in `ContractOp::regionBuilder(...)`.

Reply:
Thanks for the reference! 

From what I can see, we didn't really "explore" that thread. In my mental model (and for consistency with `vector.contract`), we should only really control the combining Op.

This is something that we can totally resolve later, but it would be good to document this.

Reply:
I have added a TODO to the docs now.

After this PR lands I will make another (small!) PR ASAP to enable control over the combining op. Those concerned about that can then chime in, e.g. @ftynse.